lib/AST/DiagnosticEngine.cpp

```diff
@@ -1016,13 +1016,14 @@ const char *DiagnosticEngine::diagnosticStringFor(const DiagID id,
   if (printDiagnosticName) {
     return debugDiagnosticStrings[(unsigned)id];
   }
-  auto defaultMessage = diagnosticStrings[(unsigned)id];
+  auto defaultMessage = defaultLocalization.get()->getMessageOr(id);
```

Comment:
This whole method needs to be rewritten now to first check whether it's possible to produce a localized message and only afterwards access `defaultLocalization` to avoid double reads when it's not necessary.